From: Helmut Buchsbaum <helmut.buchsbaum@opensource.tttech-industrial.com>
Date: Wed, 17 Jul 2019 19:42:44 +0200
Subject: dm: acpi: validate minimal length of PSDS

Kontron COMe-mAL10 does not provide an ACPI PSDS table as expected.
basl_fwrite_psds() expects to copy 33 bytes beginning at offset 46, thus
accessing the table up to offset 79. The PSDS table provided here is
shorter and thus should be ignored. Introduce a minimum length check
to achieve this.

Tracked-On: #4846
Signed-off-by: Helmut Buchsbaum <helmut.buchsbaum@opensource.tttech-industrial.com>
---
 devicemodel/hw/platform/acpi/acpi.c | 18 +++++++++++++++---
 1 file changed, 15 insertions(+), 3 deletions(-)

diff --git a/devicemodel/hw/platform/acpi/acpi.c b/devicemodel/hw/platform/acpi/acpi.c
index 3d81597..1512ad5 100644
--- a/devicemodel/hw/platform/acpi/acpi.c
+++ b/devicemodel/hw/platform/acpi/acpi.c
@@ -1200,6 +1200,7 @@ acpi_build(struct vmctx *ctx, int ncpu)
 {
 	int err;
 	int i;
+	struct stat psds_stat;
 
 	basl_ncpu = ncpu;
 
@@ -1221,9 +1222,20 @@ acpi_build(struct vmctx *ctx, int ncpu)
 	err = basl_make_templates();
 
 	/* Check PSDS in SOS, only present PSDS table to UOS when PSDS present in SOS */
-	psds_fd = open("/sys/firmware/acpi/tables/PSDS", O_RDONLY);
-	if (psds_fd >=0)
-		acpi_table_enable(PSDS_ENTRY_NO);
+	err = stat("/sys/firmware/acpi/tables/PSDS", &psds_stat);
+	if ((err == 0) && (psds_stat.st_size >= 79)) {
+		psds_fd = open("/sys/firmware/acpi/tables/PSDS", O_RDONLY);
+		if (psds_fd >= 0)
+			acpi_table_enable(PSDS_ENTRY_NO);
+	} else {
+		if (err) {
+			fprintf(stderr, "ACPI PSDS table not available, skipping\n");
+		} else {
+			fprintf(stderr, "ACPI PSDS table too small (%lu < 79), skipping\n",
+				psds_stat.st_size);
+		}
+		err = 0;
+	}
 
 	/*
 	 * Run through all the ASL files, compiling them and
